### PR TITLE
Sign container images

### LINKED ISSUE #19282: Image Signing
## Proposal / RFE

**Is your feature request related to a problem?**
Kubernetes itself is currently adding support for signing release artifacts (including cluster component images) using [cosign](https://github.com/sigstore/cosign). It would be excellent if Cilium could sign their release images as well during the build pipeline so that those users desiring more secure clusters could use an admission controller (like [connaisseur](https://github.com/sse-secure-systems/connaisseur) that validates image signatures as an additional layer of security. 

Connaisseur allows for enabling verification for whitelisted namespaces and because Cilium is typically deployed into the kube-system namespace it would enable enabling verification for the kube-system namespace once the Kubernetes images are signed as well.

https://github.com/kubernetes/enhancements/issues/3031
https://github.com/kubernetes/release/issues/2227
https://github.com/kubernetes/release/issues/2383

https://github.com/sigstore/cosign
https://github.com/sse-secure-systems/connaisseur

### PR DESCRIPTION
**Pull request by sandipanpanda**
Implement container image signing using cosign in build-images workflows. Leveraging image signing gives users confidence that the container images they got from the container registry were the trusted code that the maintainer built and published.

Fixes: #19282

Signed-off-by: Sandipan Panda <samparksandipan@gmail.com>

```release-note
Sign container images with cosign
```

### CODE REVIEW

**Comment by sandipanpanda** [comment]
CI run to test signing container image using cosign in https://github.com/cilium/cilium/actions/runs/3228306057/jobs/5284228087

**Review by nbusseneau — CHANGES_REQUESTED** [review]
You now have multiple PRs doing similar things:

- #21374
- #21611
- #21739

Can you please consolidate these to remove duplication / conflicts, and make reviewing easier? I would propose to use a single PR with everything, and have multiple commits dedicated to single things (e.g. one for signing, one for SBOM, etc.). This will be easier to handle than conflicting PRs :)

**Comment by sandipanpanda** [comment]
The purpose of the other two PRs is to test container image signing and SBOM generation by implementing the features first in build-images-ci.yaml and then opening a PR from a branch inside the Cilium repository to run the GH workflow with the changes. After the testing is successful, these features can be implemented in the build-images workflows as per the discussions with my mentors. As there are two separate issues for container image signing and SBOM generation, I thought it might be better to make separate PRs to implement those features. Adding everything into a single PR would require a lot of changes while opening a PR from a branch inside the Cilium repository to run the GH workflow with the changes.

**Comment by sandipanpanda** [comment]
cc @qmonnet. While re-requesting review from @nbusseneau, review request from @qmonnet was automatically removed.

File: Documentation/configuration/index.rst
Comment (qmonnet) [inline Documentation/configuration/index.rst]: Please let's move this section to a dedicated file, and only add a link to it in the `index.rst`.

File: Documentation/configuration/index.rst
Comment (qmonnet) [inline Documentation/configuration/index.rst]:
```suggestion
    ``COSIGN_EXPERIMENTAL=1`` is used to allow verification of images signed in 
    ``KEYLESS`` mode. To learn more about keyless signing, please refer to `Keyless Signatures`_.
```

File: Documentation/configuration/index.rst
Comment (qmonnet) [inline Documentation/configuration/index.rst:45]: Thanks for mentioning what `COSIGN_EXPERIMENTAL` is for. Further question, do you know the `_EXPERIMENTAL` suffix implies in terms of lifetime for the feature? Is this something that is not completely stable yet, or is it an experiment that could be removed from `cosign` in a future version by its maintainers if not deemed interesting?

**Review by qmonnet — CHANGES_REQUESTED** [review]
Hi and thanks!

I have no idea how `cosign` works, so I haven't validated the technical steps. Nicolas I'm leaving it to you, let me know if you need a second look on that.

I've got a few comments on the docs, see below.

Also one question on `COSIGN_EXPERIMENTAL`. Thanks for mentioning what the option is for. Do you know what the `_EXPERIMENTAL` suffix implies in terms of lifetime for the keyless feature? Is this something that is not completely stable yet, or is it an experiment that could be removed from cosign in a future version by its maintainers if not deemed interesting?

**Review by nbusseneau — APPROVED** [review]
LGTM for the workflows part.

**Review by qmonnet — APPROVED** [review]
Thanks for addressing the comments.

> but they are working on stabilizing it.

And thanks for looking into this! OK, all good then

**Review by aanm — APPROVED** [review]

**Comment by nbusseneau** [comment]
Example of a failure after merge: https://github.com/cilium/cilium/actions/runs/3298535923/jobs/5440783983

Not sure what changed between previous tests and the merge, we'll have to double check.

File: Documentation/configuration/verify-image-signatures.rst
Comment (asraa) [inline Documentation/configuration/verify-image-signatures.rst:29]: Hi! Just chiming in. Great work! I saw Dan tweet about it earlier.

Just FYI: this verification checks if there is *ANY* valid signature on the image, but not who or what signed it to determine it's trust. This means that anyone (with write access to the container) could have signed it in order to produce a positive verification, but *not* necssarily the cilium repository or maintainers.

See https://github.com/sigstore/cosign/issues/2056

I would recommend adding in a `--cert-identity .github/workflows/build-images-hotfixes.yaml` or something similar, that would allow a user to validate that the cilium repository signed the image. You could also sign your images with a service account email that the workflow authenticates to. Feel free to tag me if you're interested in something like that!

**Comment by sandipanpanda** [comment]
> Example of a failure after merge: https://github.com/cilium/cilium/actions/runs/3298535923/jobs/5440783983
> 
> Not sure what changed between previous tests and the merge, we'll have to double check.

@nbusseneau thanks to @aanm for catching it, we have to add `if: ${{ github.event_name != 'pull_request_target' }}` for # master branch pushes and `if: ${{ github.event_name == 'pull_request_target' }}`for # PR updates in the `- name: Sign CI Images` steps in [build-images-ci.yaml](https://github.com/cilium/cilium/blob/master/.github/workflows/build-images-ci.yaml)

File: .github/workflows/build-images-ci.yaml
Comment (asraa) [inline .github/workflows/build-images-ci.yaml:196]: it seems like the failure is due to missing shas here?

**Comment by sandipanpanda** [comment]
Additionally, we have to add `if: ${{ github.event_name != 'pull_request_target' }}` for # master branch pushes and `if: ${{ github.event_name == 'pull_request_target' }}`for # PR updates in the `- name: Sign CI Images` steps in [build-images-ci.yaml](https://github.com/cilium/cilium/blob/master/.github/workflows/build-images-ci.yaml)

**Comment by sandipanpanda** [comment]
Update: Implemented the feature to sign Cilium images using cosign with the required changes addressed.

**Pull request merged**
